### ISSUE TITLE
dev-lang/ghc-9.0.2-r2 does not compile on Musl (no lib64)

### ISSUE THREAD

**Issue by Thymelizabeth** [gentoo-haskell/gentoo-haskell#1368]
[build.log](https://github.com/gentoo-haskell/gentoo-haskell/files/10147906/build.log)
[emerge-info.txt](https://github.com/gentoo-haskell/gentoo-haskell/files/10147909/emerge-info.txt)

Fails in the prepare phase trying to relocate folders.

**Comment by hololeap** [comment]
Here's are some relevant blocks of code:

https://github.com/gentoo-haskell/gentoo-haskell/blob/866a3775a18e5e8459fc2e9bb6982d0bfbe91525/dev-lang/ghc/ghc-9.0.2-r2.ebuild#L436-L444

https://github.com/gentoo-haskell/gentoo-haskell/blob/866a3775a18e5e8459fc2e9bb6982d0bfbe91525/dev-lang/ghc/ghc-9.0.2-r2.ebuild#L292-L299

https://github.com/gentoo-haskell/gentoo-haskell/blob/866a3775a18e5e8459fc2e9bb6982d0bfbe91525/dev-lang/ghc/ghc-9.0.2-r2.ebuild#L305-L320

I take it you are using the `binary` USE flag?

**Comment by hololeap** [comment]
> dev-lang/ghc-9.0.2-r2 does not compile on Musl (no lib64)

@Thymelizabeth are you aware that using the `binary` USE flag skips compilation, and instead just installs directly from the binary we provide?

https://github.com/gentoo-haskell/gentoo-haskell/blob/4813a4bcd59b3b618199e2ed8a6eecf1a7d1ef1e/dev-lang/ghc/metadata.xml#L9

**Comment by Thymelizabeth** [comment]
The way I see it, there are several problems with trying to install GHC on a system using musl:

1. The only binaries downloaded are compiled for glibc. These cannot be used to build GHC and cannot be installed directly to the system. Despite this, Portage sees no problem trying to install GHC on a musl based system. There is a build for Alpine Linux available from [the main Haskell website](https://www.haskell.org/ghc/download.html) which could be added. I recently changed the [dev-dotnet/dotnet-sdk-bin package](https://gitweb.gentoo.org/repo/gentoo.git/tree/dev-dotnet/dotnet-sdk-bin/dotnet-sdk-bin-7.0.100-r1.ebuild) to change the binary download depending on the `libc` the system uses. This ebuild could use something similar.

2. When **_not_** using the `binary` USE flag, Portage tries to rename a non-existent folder, i.e. the `/usr/lib` folder to `/usr/lib64`. This folder is not provided in the binary. It is, however, provided in the source. `/usr/lib64` already exists in the binary. [musl does not recommend using a `lib64` or `lib32` directory](https://wiki.musl-libc.org/guidelines-for-distributions.html)   

3. When trying to install **_with_** the `binary` USE flag, Portage tries to backup the `/usr/lib/ghc-9.0.2/package.conf.d` directory. This folder is also not provided in the binary package, and is not produced as the ebuild is processed. Bear in mind, this binary package is not compiled for musl, but Portage sees nothing wrong with trying to install it.

**Comment by hololeap** [comment]
@Thymelizabeth Do you think that something like this would be a good first step, just to ensure people on `musl` don't try to download our glibc binary?

```diff
diff --git a/dev-lang/ghc/ghc-9.0.2-r2.ebuild b/dev-lang/ghc/ghc-9.0.2-r2.ebuild
index 1bbe52e971..9e6149913c 100644
--- a/dev-lang/ghc/ghc-9.0.2-r2.ebuild
+++ b/dev-lang/ghc/ghc-9.0.2-r2.ebuild
@@ -29,7 +29,8 @@ BIN_PV=${PV}
 #arch_binaries="$arch_binaries alpha? ( https://slyfox.uni.cx/~slyfox/distfiles/ghc-bin-${PV}-alpha.tbz2 )"
 #arch_binaries="$arch_binaries arm? ( https://slyfox.uni.cx/~slyfox/distfiles/ghc-bin-${PV}-armv7a-hardfloat-linux-gnueabi.tbz2 )"
 #arch_binaries="$arch_binaries arm64? ( https://slyfox.uni.cx/~slyfox/distfiles/ghc-bin-${PV}-aarch64-unknown-linux-gnu.tbz2 )"
-arch_binaries="$arch_binaries amd64? ( https://eidetic.codes/ghc-bin-${PV}-x86_64-pc-linux-gnu-r1.tbz2 )"
+arch_binaries="$arch_binaries amd64? ( elibc_glibc? ( https://eidetic.codes/ghc-bin-${PV}-x86_64-pc-linux-gnu-r1.tbz2 ) )"
+#arch_binaries="$arch_binaries amd64? ( elibc_musl? ( https://eidetic.codes/ghc-bin-${PV}-x86_64-pc-linux-musl-r1.tbz2 ) )"
 #arch_binaries="$arch_binaries ia64?  ( https://slyfox.uni.cx/~slyfox/distfiles/ghc-bin-${PV}-ia64-fixed-fiw.tbz2 )"
 #arch_binaries="$arch_binaries ppc? ( https://slyfox.uni.cx/~slyfox/distfiles/ghc-bin-${PV}-ppc.tbz2 )"
 #arch_binaries="$arch_binaries ppc64? ( https://slyfox.uni.cx/~slyfox/distfiles/ghc-bin-${PV}-ppc64.tbz2 )"
@@ -46,7 +47,15 @@ yet_binary() {
                #alpha) return 0 ;;
                #arm64) return 0 ;;
                #arm) return 0 ;;
-               amd64) return 0 ;;
+               amd64)
+                       if use elibc_glibc; then
+                               return 0
+                       elif use elibc_musl; then
+                               return 1
+                       else
+                               return 1
+                       fi
+                       ;;
                #ia64) return 0 ;;
                #ppc) return 0 ;;
                ppc64)
```

Also, have you been able to build successfully with the `ghcbootstrap` USE flag using an instance of GHC installed through e.g. `ghcup`?

**Comment by Thymelizabeth** [comment]
The patch looks good.

I tried installing with `ghcbootstrap` enabled, (using the Alpine tarball installed to `/usr/local`), but got a segmentation fault when `make` tried to use a non-existent file.
[ghcbootstrap-enabled-build.log](https://github.com/gentoo-haskell/gentoo-haskell/files/10191808/ghcbootstrap-enabled-build.log)

**Comment by thesamesam** [comment]
> I'm not sure if `amd64` and `amd64-musl` need to be treated as separate build hosts, e.g. `musl` needs its own binary.
> 
> @thesamesam might have some insight here.

Yeah, musl will always need its own set as it's a different ABI (treat it like a different architecture for the purposes of reusing binaries).

**Comment by thesamesam** [comment]
> GHCup also segfaults. [ghcup-make.log](https://github.com/gentoo-haskell/gentoo-haskell/files/10192723/ghc-make.log) Unrelated issue: [gold (binutils) has a bug](https://sourceware.org/bugzilla/show_bug.cgi?id=23856) where dynamically linked executables with musl segfault on startup, so I used lld with ghcup.

I would not use gold for anything, it's considered deprecated.

**Comment by formula-spectre** [comment]
I suggest taking a look here:
https://github.com/clang-musl-overlay/clang-musl-overlay/tree/master/dev-lang/ghc
I use this ebuild on musl and it works fine
